File: python/python-sdk/kestrapy/api/kv_api.py
from typing import List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.kv_controller_api_delete_bulk_request import KVControllerApiDeleteBulkRequest
from kestrapy.models.kv_controller_api_delete_bulk_response import KVControllerApiDeleteBulkResponse
from kestrapy.models.kv_controller_kv_detail import KVControllerKvDetail
from kestrapy.models.kv_entry import KVEntry
from kestrapy.models.paged_results_kv_entry import PagedResultsKVEntry
from kestrapy.models.query_filter import QueryFilter


class KVApi(BaseApi):

    # ── Key-Value CRUD ──────────────────────────────────────────────

    def set_key_value(
        self,
        namespace: str,
        key: str,
        tenant: str,
        body: str,
    ) -> None:
        path = self._tenant_path(tenant, "namespaces", namespace, "kv", key)
        self._void_request("PUT", path, body=body, content_type=self.TEXT)

    def key_value(
        self,
        namespace: str,
        key: str,
        tenant: str,
    ) -> KVControllerKvDetail:
        path = self._tenant_path(tenant, "namespaces", namespace, "kv", key)
        return self._json_request("GET", path, KVControllerKvDetail)

    def delete_key_value(
        self,
        namespace: str,
        key: str,
        tenant: str,
    ) -> bool:
        path = self._tenant_path(tenant, "namespaces", namespace, "kv", key)
        return self._raw_json_request("DELETE", path)

    def delete_key_values(
        self,
        namespace: str,
        tenant: str,
        request: KVControllerApiDeleteBulkRequest,
    ) -> KVControllerApiDeleteBulkResponse:
        path = self._tenant_path(tenant, "namespaces", namespace, "kv")
        return self._json_request("DELETE", path, KVControllerApiDeleteBulkResponse, body=request)

    # ── List & Search ───────────────────────────────────────────────

    def list_all_keys(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> PagedResultsKVEntry:
        path = self._tenant_path(tenant, "kv")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsKVEntry, params=params)

    def list_keys_with_inheritance(
        self,
        namespace: str,
        tenant: str,
    ) -> List[KVEntry]:
        path = self._tenant_path(tenant, "namespaces", namespace, "kv", "inheritance")
        return self._json_list_request("GET", path, KVEntry)

File: go-sdk/kestra_api_client/kv_api.go
package kestra_api_client

import (
	"context"
	"net/url"
)

type KvAPI struct {
	baseAPI
}

func (a *KvAPI) SetKeyValue(ctx context.Context, namespace, key, tenant, body string) error {
	return a.doVoidText(ctx, "PUT", tenantPath(tenant, "namespaces", namespace, "kv", key), body, nil)
}

func (a *KvAPI) KeyValue(ctx context.Context, namespace, key, tenant string) (*KvDetail, error) {
	return doJSON[*KvDetail](&a.baseAPI, ctx, "GET", tenantPath(tenant, "namespaces", namespace, "kv", key), nil, nil)
}

func (a *KvAPI) DeleteKeyValue(ctx context.Context, namespace, key, tenant string) (*bool, error) {
	return doJSON[*bool](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "namespaces", namespace, "kv", key), nil, nil)
}

func (a *KvAPI) DeleteKeyValues(ctx context.Context, namespace, tenant string, request KVControllerApiDeleteBulkRequest) (*KVControllerApiDeleteBulkResponse, error) {
	return doJSON[*KVControllerApiDeleteBulkResponse](&a.baseAPI, ctx, "DELETE", tenantPath(tenant, "namespaces", namespace, "kv"), request, nil)
}

func (a *KvAPI) ListAllKeys(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter) (*PagedResultsKVEntry, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsKVEntry](&a.baseAPI, ctx, "GET", tenantPath(tenant, "kv"), nil, params)
}

func (a *KvAPI) ListKeysWithInheritance(ctx context.Context, namespace, tenant string) ([]KVEntry, error) {
	return doJSON[[]KVEntry](&a.baseAPI, ctx, "GET", tenantPath(tenant, "namespaces", namespace, "kv", "inheritance"), nil, nil)
}

func (a *KvAPI) SetKeyValueWithTTL(ctx context.Context, namespace, key, tenant, body string, ttl *string) error {
	var params url.Values
	if ttl != nil {
		params = buildQueryParams("ttl", ttl)
	}
	return a.doVoidText(ctx, "PUT", tenantPath(tenant, "namespaces", namespace, "kv", key), body, params)
}
